src/journal/compress.c

```diff
@@ -541,6 +645,90 @@ int compress_stream_lz4(int fdf, int fdt, uint64_t max_bytes) {
 #endif
 }
 
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *s = ZSTD_createCStream();
+        size_t out_size = ZSTD_CStreamOutSize();
+        uint8_t out[out_size];
+        char *src = NULL;
+        struct stat st;
+        ZSTD_inBuffer ZSTD_in = {src, 0, 0};
+        ZSTD_outBuffer ZSTD_out = {&out, sizeof(out), 0};
+        size_t ret;
+        uint64_t left = max_bytes;
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        if (!s)
+                return -ENOMEM;
+
+        ret = ZSTD_initCStream(s, ZSTD_COMPRESSION_LEVEL);
+        if (ZSTD_isError(ret)) {
+                log_error("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errnoh(ret);
+        }
+
+        if (fstat(fdf, &st) < 0)
+                return log_debug_errno(errno, "fstat() failed: %m");
+        ZSTD_in.size = st.st_size;
+
+        src = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fdf, 0);
```

Comment:
i know the lz4 code does this, but i am not sure the mmap() thing is a good idea for this kind of stuff... for stream access mmap is usually slower than read()/write() (since trapping in memory is actually slow), and this means we can user the function only on regular files, not on anything actually stream based, i.e. a socket or pipe. The function is a bit of a misnomer hence by using mmap()... Moreover it means we can only process files that are mappable as a whole. Which coredumps might be, but this might fall apart as soon as we map anything else larger in parallel. Now, we only use it on regular files, hence we get away with using mmap(), but it am pretty sure this is not worth it, and for new code we should probably use read(), unless it really can be shown that mmap()ing makes it faster...

Reply:
We can't bail out early when max_bytes is too small unless we do it this way, if we use read() we have to read to decompress up to max_bytes before we know it won't fit.

Reply:
`ZSTD_findDecompressedSize()` is used when you might have 2 zstd compressed frames concatenated together, and is part of the unstable API. If that isn't the case, you can use `ZSTD_getFrameContentSize()`, which only needs to look at up to the first 15 bytes, so we can use `read()` and `write()`.